File: tree/257/c++/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        vector<int> path;
        solve(res, root, path);
        return res;
    }
    
    void solve(vector<string>&res, TreeNode* root, vector<int> path){
        if(!root)
            return;
        path.push_back(root->val);
        if(!root->left && !root->right){
            string s = "";
            for(int i=0; i< path.size()-1;i++){
                s += to_string(path[i]) + "->";
            }
            s += to_string(path[path.size()-1]);
            res.push_back(s);
        }
        solve(res,root->left,path);
        solve(res,root->right,path);
    }
};

File: tree/257/java/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 * DFS遍历
 */
public class Solution {
   
    public List<String> binaryTreePaths(TreeNode root) {
        if(root == null){
            return Collections.emptyList();
        }
        List<String> res = new ArrayList<>();
        Deque<TreeNode> st = new ArrayDeque();
        Deque<TreeNode> path = new ArrayDeque();
        TreeNode t;
        st.addFirst(root);
        while(!st.isEmpty()){
            t=st.removeFirst();
            if(t!=path.peekLast()){
                path.addLast(t);
                st.addFirst(t);
            }
            else{
                path.removeLast();
                continue;
            }
            if(t.right==null&&t.left==null){
                Iterator<TreeNode> it = path.iterator();
                StringBuilder sb = new StringBuilder();
                sb.append(Integer.toString(it.next().val));
                while(it.hasNext()){
                    sb.append("->");
                    sb.append(Integer.toString(it.next().val));
                }
                res.add(sb.toString());
            }
            if(t.right!=null){
                st.addFirst(t.right);
            }
            if(t.left!=null){
                st.addFirst(t.left);
            }
        }
        return res;
    }
}
